File: Data Structures & Algorithms/binary-search/submission-4.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 1:
            return 0 if nums[0] == target else -1

        l, r = 0, len(nums) - 1

        while l <= r:
            mid = (l+r)//2

            num = nums[mid]

            if num == target:
                return mid
            
            if num > target:
                r -= 1
            else:
                l += 1
        
            
        return -1
        

File: Data Structures & Algorithms/top-k-elements-in-list/submission-4.rs
use std::collections::{BinaryHeap, HashMap};

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, mut k: i32) -> Vec<i32> {
        let mut counter = HashMap::<i32, usize>::new();

        for num in nums {
            *counter
            .entry(num)
            .or_default() += 1
        }

        let mut heap: BinaryHeap<(usize, i32)> = counter
                                            .iter()
                                            .map(|(&num, &count)| (count, num))
                                            .collect();
        
        let mut res = Vec::<i32>::new();
        while k > 0 {
            if let Some((_, num)) = heap.pop() {
                res.push(num);
                k -= 1;
            }
        }

        res
    }
}
